src/main/java/com/google/api/generator/gapic/protoparser/Parser.java

```diff
@@ -641,10 +640,51 @@ static LongrunningOperation parseLro(
     OperationInfo lroInfo =
         methodDescriptor.getOptions().getExtension(OperationsProto.operationInfo);
 
-    String responseTypeName = parseNestedProtoTypeName(lroInfo.getResponseType());
-    String metadataTypeName = parseNestedProtoTypeName(lroInfo.getMetadataType());
-    Message responseMessage = messageTypes.get(responseTypeName);
-    Message metadataMessage = messageTypes.get(metadataTypeName);
+    // These can be short names (e.g. FooMessage) or fully-qualified names with the *proto* package.
+    String responseTypeName = lroInfo.getResponseType();
+    String metadataTypeName = lroInfo.getMetadataType();
+    String[] responseTypeNameComponents = responseTypeName.split("\\.");
+    String[] metadataTypeNameComponents = metadataTypeName.split("\\.");
+    String responseTypeShortName =
+        responseTypeNameComponents[responseTypeNameComponents.length - 1];
+    String metadataTypeShortName =
+        metadataTypeNameComponents[metadataTypeNameComponents.length - 1];
+
+    // True if the LRO-specified name is the short name.
+    boolean isResponseTypeNameShortOnly = responseTypeName.equals(responseTypeShortName);
+    boolean isMetadataTypeNameShortOnly = metadataTypeName.equals(metadataTypeShortName);
+
+    Message responseMessage = null;
+    Message metadataMessage = null;
+
+    // The messageTypes map keys to the Java fully-qualified name.
+    for (Map.Entry<String, Message> messageEntry : messageTypes.entrySet()) {
+      String[] packageComponents = messageEntry.getKey().split("\\.");
+      String messageShortName = packageComponents[packageComponents.length - 1];
+      if (responseMessage == null) {
```

Comment:
The `responseMessage` logic and `metadataMessage` logic seem to be a copypaste of each other. Please consider moving that into a private or method combining this to a for loop with 2 elements in it (response and metadata messages).

Reply:
I'd considered using these, which would result in the following tradeoffs:
- Nested loop: Need an extra map or data structure to associate `responseMessage`, `responseTypeName`, `isResponseTypeNameShortOnly`, which would be less readable, or recompute the latter in the loop (which is inefficient).
- Separate method: Need to take four parameters at best - the three above plus `messageEntry`, which would make the call site, helper signature, and method itself less readable.

Given that the logic is relatively small, copy-pasting does not seem to have significantly better or worse tradeoffs compared to the alternative approches.